refactor: Transition from `useAsync` to RTK Query (M2-8618)

farmerpaul opened the pull request:
<!-- Use this template as a guide to describe your pull request, and adjust as necessary. -->
<!-- Include information that helps your peers review your updates and understand this    -->
<!-- repository's history of changes over time.                                           -->

<!-- Delete any options that are not relevant -->

- [x] Tests for the changes have been added
- [x] Related documentation has been added / updated

### 📝 Description

> [!IMPORTANT]
> As this is a chunky PR, I took time to meticulously organize the commits into related changes, along with detailed documentation in the form of commit messages, with the hopes that this will help reduce friction for reviewers. Thus, it's **highly recommended** to review [commit by commit](https://github.com/ChildMindInstitute/mindlogger-admin/pull/2021/commits).

🔗 Jira Ticket [M2-8618](https://mindlogger.atlassian.net/browse/M2-8618)

This PR begins the migration from `useAsync` and raw Axios mutation functions to [RTK Query](https://redux-toolkit.js.org/rtk-query/overview), which has been agreed upon as the new paradigm for making API calls in the Admin App.

See this Confluence doc for context:

- [**Improving Endpoint Caching in Admin App**](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/1036058625/Improving+Endpoint+Caching+in+Admin+App)

In this PR, these endpoints have been fully migrated over to RTK Query, as a sensible starting point for the ongoing overall migration to this framework:

- GET `activities/applet/${appletId}`
- GET `workspaces/${ownerId}/managers`
- GET `workspaces/${ownerId}/applets/${appletId}/managers`
- GET `workspaces/${ownerId}/respondents`
- GET `workspaces/${ownerId}/applets/${appletId}/respondents`
- POST `invitations/${appletId}/${url}`
- POST `invitations/${appletId}/subject`
- POST `invitations/${appletId}/shell-account`
- PUT `subjects/${subjectId}`
- DELETE `subjects/${subjectId}`

#### 👀 Other things worth noting:

- A [custom base query function](https://github.com/ChildMindInstitute/mindlogger-admin/blob/refactor/rtk-query/src/shared/api/apiSlice.ts#L9), `axiosBaseQuery`, has been written, allowing RTK Query to use `axios` instead of `fetch` for its HTTP requests (which it uses by default). Continuing to use `axios` for network requests allows unit tests to remain mostly unaffected by the migrated endpoints (preserving the use of `jest-mock-axios`), as well as the existing authentication stack to be used with no changes needed.
- The Redux infrastructure for `allRespondents`, which has acted as a rudimentary "cache" for participants used by a handful of components, is no longer needed with the migration to RTK Query, as the new querying infra provides the same functionality but more reliably and with more intuitive caching. So the infra related to that portion of Redux state has been removed.

### 🪤 Peer Testing

The app should remain **functionally the same** as before these changes, with the exception of improved performance (demoed <a href="https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/1036058625/Improving+Endpoint+Caching+in+Admin+App#Fix-duplicated-API-Calls" rel="nofollow">here</a>) and a couple minor UX bugs fixed (demoed <a href="https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/1036058625/Improving+Endpoint+Caching+in+Admin+App#Fix-data-becoming-stale%2Fout-of-sync" rel="nofollow">here</a>). So peer testing should be done with the intent of **verifying there are no regressions**.

Below are the screens and components affected by the migrated queries and mutations. When testing, ensure that queried data – including lists of activities, participants, and team members – remains up to date when performing actions that would affect that data.

| Original API function | Affected screens/actions |
| - | - |
| `getAppletActivitiesApi` | <ul><li>**Applet > Activities** list</li><li>**Assign Activity** drawer (accessed either from Applet > Activities or PDP)</li></ul> |
| `getWorkspaceManagersApi` | <ul><li>**Dashboard > Team** table</li><li>**Applet > Team** table</li></ul> |
| `getWorkspaceRespondentsApi` | <ul><li>**Dashboard > Respondents** table</li><li>**Applet > Participants** table</li><li>**Participant dropdowns**, including in:<ul><li>**Take Now** modal</li><li>**Assign Activity** drawer</li></ul></li><li>**Applet > Team** screen: <ul><li>**Add Team Member** > **Reviewer** role > **Select Participants** dropdown</li><li>**Edit Team Member** on "Reviewer" > **Select Respondents** modal</li></ul></li><li>**Applet > Participants > PDP > Schedule**: <ul> <li> **Add individual schedule** </li> <li> **Remove individual schedule** <br>☝️ Confirm page state reflects individual schedule state </li> </ul></li></ul> |
| `postAppletInvitationApi`<br>`postAppletShellAccountApi`<br>`postSubjectInvitationApi` | <ul><li>**Dashboard > Respondents** screen: **Send Invitation** action on limited accounts</li><li>**Applet Overview** screen: **⊕ Add Participant** action</li><li>**Applet > Participants** screen: <ul><li>**Add Participant** action, both limited & full accounts</li><li>**Upgrade to Full Account** action on limited accounts</li></ul></li><li>**Applet > Team** screen: **Add Team Member** action</li></ul> |
| `editSubjectApi`<br>`deleteSubjectApi` | <ul><li>**Applet > Participants** screen:<ul><li>**Edit Participant** action</li><li>**Remove Participant** action</li></ul></li></ul> |

[M2-8618]: https://mindlogger.atlassian.net/browse/M2-8618?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

farmerpaul committed:
feat: add RTK Query core infra

Create RTK Query adapter to use Axios instead of fetch, allowing us to
piggyback on existing auth/API infrastructure.

farmerpaul committed:
refactor: adapt API error handling for RTK Query

To prevent Redux console warnings, we need to use a serializable error
type for `axiosBaseQuery`. It made sense to reuse the default error type
for RTK Query, `FetchBaseQueryError`. However, this means adapting our
error handling utilities to support this type, so made those changes in
this commit.

farmerpaul committed:
refactor: Dashboard user/subject endpoints ⇒ RTKQ

Migrate Dashboard module endpoints that do CRUD of respondent, manager,
and subject entities and fetching of applet activities to a new
RTK Query slice that lives in the Dashboard module.

Also clean up some API types:
- `ParticipantsData` represents an API response, so define it alongside
existing API response types aligning with existing naming convention
(`WorkspaceRespondentsResponse`)
- Same with `ManagersData` => `WorkspaceManagersResponse`
- Add some missing fields to `EditSubjectResponse`
- Restrict `url` param in `AppletInvitationData` type to valid options

farmerpaul committed:
refactor: `getAppletActivitiesApi` calls ⇒ RTKQ

farmerpaul committed:
refactor: `getWorkspaceManagerApi` calls ⇒ RTKQ

farmerpaul committed:
refactor: edit/remove subject mutations ⇒ RTKQ

RTK Query doesn't use "on success" callback as `useAsync` does, so the
business logic for responding after the mutation can be kept together.
We'll see the same pattern in future commits as well.

farmerpaul committed:
refactor: create full/limited/team accounts ⇒ RTKQ

Migrate mutation calls for creating account invitations and limited
accounts to RTK Query. (Also remove legacy edit/delete subject mutation
functions, which I forgot to include in the previous commit.)

Popups associated with each of these mutations no longer need to inform
the consuming component that a refetch is needed, since that's handled
automatically by RTK Query's cache invalidation system.

It's worth noting that the Applet Overview screen was also responding to
the `shouldRefetch` flag, refreshing the list of submissions in case a
new participant is added. However, submissions are not immediately
affected by newly added participants, so that refetch was also
superfluous.

farmerpaul committed:
refactor: `getWorkspaceRespondentsApi` ⇒ RTKQ [#1]

Migrate calls made to `getWorkspaceRespondentsApi` to RTK Query, part 1:
Migrate calls from Respondents and Applet > Participants screens as well
as all participant dropdowns.

In making this change, also added support for missing loading states for
pinning/unpinning participants.

farmerpaul committed:
refactor: `getWorkspaceRespondentsApi` ⇒ RTKQ [#2]

Migrate remaining `getWorkspaceRespondentsApi` calls to RTK Query:
Migrate references to the `useAllRespondentsData` and
`getAllWorkspaceRespondents` functions provided by Redux infra, which
are used to "cache" participant data in Redux state and allow updating
it via a thunk (where the only remaining call to
`getWorkspaceRespondentsApi` still lives). Instead of using this awkward
Redux infra, we'll use RTK Query directly.

Consumers of this infra were using it for various purposes, often
inefficiently:
- `AddManagerPopup` and `SelectRespondentsPopup` were using it to fetch
a list of all participants; migrating these to RTK Query will allow
better harnessing of cache.
- Applet > Schedule was using it solely for a superfluous permission
check, probably from before that screen was refactored to become
participant-agnostic as of multi-informant changes last year.
- in Applet > Participants > PDP > Schedule was using it to get only
the currently viewed participant's schedule information, specifically
their `hasIndividualSchedule` flag. Made this more efficient by
employing a more targeted respondents query that filters the fetch on
the BE by the current participant, rather than doing that client-side.
- Other components were using the infra effectively to invalidate the
"cache" at specific times. Added comments to these spots including TODOs
to remove this manual step once other mutations are migrated to RTKQ.

farmerpaul committed:
chore: remove unneeded Redux infra

Removed `getWorkspaceRespondentsApi` and its last instance of use in
the Redux infra for `allRespondents` data. That Redux clutter is now no
longer in use.

farmerpaul committed:
chore: update tests

With the changes in this PR, a number of tests needed tweaking to pass:

- The `WorkspaceRespondentsResponse` type becoming more restrictive
(including the new data access info) required updating participant mocks
and associated tests.
- The `allRespondents` preloaded state is no longer used and needed to
be cleaned up. In cases where we now are using RTK Query to return that
data, some of the test logic required tweaks to wait for the mock
fetches to complete, since the data is no longer preloaded in Redux.
- `mockAxios` calls no longer always have `{ signal: undefined }`
passed to them, so that needs to be taken into account where those calls
were made by RTK Query.

farmerpaul committed:
doc: deprecate `useAsync`, update README

farmerpaul committed:
fix: styling of participant Pin buttons

During testing, discovered a random bug with styling of Pin buttons on
the Respondents & Applet > Participants screens. This is not a recent
regression, but a regression that may have ocurred 7 months ago and
never ended up being reported.

farmerpaul committed:
fix: eliminate TakeNowModal rerender during search

Using RTK Query can have the side-effect of extra renders if not done
carefully. Normally these would be inconsequential, however a component
that does not handle re-renders gracefully is `TakeNowModal`:

This component is redefined (recreated) each time its enclosing hook,
`useTakeNow`, has to rerun due to values changed in its own state or
called hook return values. This turns out to be a bad pattern: When the
`TakeNowModal` component is recreated, it disrupts the user experience
by hiding and reappearing along with its background overlay, and when it
reappears, its form state has been completely reset. This means that if
typing into a field (say, during search) causes `useTakeNow` to rerun,
you see the modal violently flicker, your search cancelled and form
state reset.

Although the best solution would be to refactor `useTakeNow` and
`TakeNowModal` so that the latter isn't recreated inside the former,
thereby allowing it to gracefully handle its parent hook's change of
state, I opted instead to work around the issue by preventing the
re-renders in the first place:

- `selectFromResult` can cause unnecessary re-renders if it doesn't use
a "selector instance", a mechanism provided by RTK Query that memoizes
the result based on inputs. See
https://redux.js.org/tutorials/essentials/part-8-rtk-query-advanced#selecting-values-from-results
Its syntax felt needlessly awkward to me, however, so I used a familiar
`useMemo` instead.
- `useLazyQuery` also causes unnecessary re-renders because it is
designed to return the query result _both_ from its returned array (the
2nd element), as well as its its execution function. Even if you discard
the 2nd array element via destructuring, that doesn't prevent the
re-render. The alternative is to use RTK Query's imperative option
involving dispatching a Redux action.

mbanting commented:
Peer testing


Original API function | Affected screens/actions
-- | --
getAppletActivitiesApi | ✅ 
getWorkspaceManagersApi | ✅ 
getWorkspaceRespondentsApi | ✅ 
postAppletInvitationApipostAppletShellAccountApipostSubjectInvitationApi | ✅ 
editSubjectApideleteSubjectApi | ✅ 

🎉

mbanting commented on src/shared/api/apiSlice.ts:
```diff
@@ -0,0 +1,99 @@
+export const apiSlice = createApi({
+  reducerPath: 'api',
+  baseQuery: axiosBaseQuery,
+  tagTypes: ['Applet', 'Subject', 'User'],
```
Convenient that RTKQ leverages TypeScript’s literal type inference to ensure only these predefined literal types are used in providesTags and invalidateTags.

mbanting commented on src/modules/Dashboard/features/Applet/Schedule/AddIndividualSchedulePopup/AddIndividualSchedulePopup.tsx:
```diff
@@ -34,11 +35,10 @@ export const AddIndividualSchedulePopup = ({
-        }),
-      );
+      // Refresh current user after creating schedule to update hasIndividualSchedule flag.
+      // TODO: When createIndividualEventsApi has been migrated to RTK Query and configured to
```
Is there a task in JIRA we can link these to?

farmerpaul replied:
Not yet, but definitely we should be creating these. I'll add one for this comment and the other 3 instances of the same comment.

mbanting commented on src/modules/Dashboard/features/Applet/Popups/AddParticipantPopup/AddParticipantPopup.types.ts:
```diff
@@ -5,7 +5,7 @@ import { UserSelectableParticipantTag } from 'shared/consts';
   popupVisible: boolean;
   appletId: string | null;
-  onClose?: (shouldRefetch: boolean) => void;
+  onClose?: () => void;
```
Nice how refetching is automatically handled now

mbanting reviewed (approved):
Monumental PR. Amazing work! In addition to a much snappier UX, it's nice that loading states and refetches are starting to be managed by the framework. You've actually reduced the # of lines of code in this repo with this PR. 

Thanks also for the readme update. 

Only a couple of minor comments, but approving already.

mbanting commented on src/shared/utils/errors.ts:
```diff
@@ -16,6 +17,10 @@ export const getErrorData = (error: ErrorData) => {
   } else if (typeof error === 'object') {
     const err = error as AxiosResponse;
 
+    if (err?.data?.result?.length) {
+      return err?.data?.result[0];
+    }
+
```
Out of curiousity, what required this addition? Was it related to RTKQ?

farmerpaul replied:
I don't think it was per se. It was just an omission I discovered as I was analyzing this code to understand better how error handling was being done. It appeared that this conditional branch was ignoring a possible error response structure (using the first `result` value), the way it's done when `error` is `instanceof Error` in the first branch. Though, maybe using the first `result` value is in fact not correct in the first place (it does seem a bit weird to me)… or maybe it's only expected when `error instanceof Error`? So perhaps rather than introduce possible regressions, I should remove these lines. What do you think?

mbanting commented on src/shared/api/apiSlice.ts:
```diff
@@ -0,0 +1,99 @@
+  reducerPath: 'api',
+  baseQuery: axiosBaseQuery,
+  tagTypes: ['Applet', 'Subject', 'User'],
+  endpoints: (builder) => ({}),
```
This was giving me an unused linter warning
```suggestion
  endpoints: (_builder) => ({}),
```

farmerpaul replied:
Funny, I was expecting a warning from this, too, but I never got one – till now. Fixed it!

mbanting commented on src/shared/hooks/useAsync/useAsync.ts:
```diff
@@ -14,6 +14,11 @@ import {
 /**
  * Utility hook to handle async operations with loading state, error state, and callbacks.
  *
+ * @deprecated This is legacy code used for handling API calls, but notably lacks caching support
```
Nice, thank you!

farmerpaul committed:
fix: PR review fixes

andrevitalb commented on src/modules/Dashboard/components/ParticipantDropdown/ParticipantDropdown.hooks.ts:
```diff
@@ -42,46 +36,26 @@ export const useParticipantDropdown = ({
+  const allParticipants = useMemo(
+    () => participants.filter(isParticipantValid).map(participantToOption),
+    [isParticipantValid, participants],
   );
```
Not 100% if this has specific implications due to the RTK implementation, but curious to know if both `useMemo` are needed here.

Would it make sense to do something like this?

```ts
  const { data: participantsData, isLoading: isFetchingParticipants } =
    useGetWorkspaceRespondentsQuery({ params: { appletId, ownerId, limit: 100 } }, { skip });
  const participants = participantsData?.result ?? [];

  const allParticipants = useMemo(
    () => participants.filter(isParticipantValid).map(participantToOption),
    [isParticipantValid, participants],
  );
```

Since the computational cost of the actual assignment of values from the fetch query would be a bit negligible.

Or even like this?


```ts
  const { data: participantsData, isLoading: isFetchingParticipants } =
    useGetWorkspaceRespondentsQuery({ params: { appletId, ownerId, limit: 100 } }, { skip });

  const allParticipants = useMemo(
    () => (participantsData ?? []).filter(isParticipantValid).map(participantToOption),
    [isParticipantValid, participants],
  );
```

farmerpaul replied:
Yup, definitely – your last version is what I went with. The first example you gave generates an ESLint warning, which is why it was wrapped in its own `useMemo` originally. But there's absolutely no need for two memos, thanks for pointing this out!

andrevitalb commented on src/modules/Dashboard/features/Applet/Activities/Activities.tsx:
```diff
@@ -27,13 +26,14 @@ export const Activities = () => {
   );
+
+  const activities: Activity[] = useMemo(() => data?.activitiesDetails ?? [], [data]);
+  const flows: ActivityFlow[] = useMemo(() => data?.appletDetail.activityFlows ?? [], [data]);
```
Similar to the other comment, do these need to live as memoized values? The actual operation for these is just the assignment. Again, this comes from a not well-versed is RTK perspective. If this is still something we'd want to keep, I'm all for it.

farmerpaul replied:
It was done to work around an ESLint warning, but I agree, it's a bit unnecessary. I was able to get rid of them and avoid the warning with a simple workaround. Thanks!

andrevitalb commented on src/modules/Dashboard/features/Applet/Overview/Overview.tsx:
```diff
@@ -52,12 +52,8 @@ export const Overview = () => {
   );
 
-  const handleAddPopupClose = (shouldRefetch = false) => {
+  const handleAddPopupClose = () => {
```
I love a good cache handling and auto refetch if required 🤓

andrevitalb commented on src/modules/Dashboard/features/Managers/Popups/SelectRespondentsPopup/SelectRespondentsPopup.test.tsx:
```diff
@@ -72,10 +65,13 @@ const successfulResponse = {
 
 describe('SelectRespondentsPopup component tests', () => {
-  test('should appear popup without selected respondents', async () => {
+  beforeEach(() => {
```
Nice QOL change

andrevitalb reviewed (approved):
Appreciate the huge amount of effort it took to get this migration rolling. Loving the simplification that comes with RTK, and app feeling way faster is a nice cherry on top. 

Awesome work, as always!

farmerpaul committed:
fix: Invalidate cache when applet updated

farmerpaul committed:
fix: Keep participants lists in sync

Ensure that when the partcipants table is refreshed, any other lists of
participants get the latest changes as well.

Also use the shared `apiSlice` whenever calling cache invalidation utils
rather than `apiDashboardSlice`.

farmerpaul committed:
chore: Reduce superfluous use of memos
